### PR TITLE
Docs: Implementation Guide Update

### PR DESCRIPTION
**Pull request by travisk1154**
## Pull Request Description
The purpose of this PR is to modify the recently added developer guide documentation so they conform with proper markdown formatting and grammatical errors. This also includes link verification and updating when needed.

## Author Checklist
* [ ] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [ ] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [ ] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

File: doc/wiki/Index.md [hunk 1]
```diff
@@ -35,6 +35,7 @@
     - [Helpful Compiler Flags](testing/Helpful_Compiler_Flags_For_Testing.md)
 
 - [Design Documentation](design/Design_Documentation_Index.md)
+- [Developer Guide](developer_guide.md)
```
Comment (hzhou) [inline doc/wiki/Index.md:38]: The link was already added in "Contributing to MPICH"

File: doc/wiki/notes/binding_c.md [hunk 2]
```diff
@@ -1,22 +1,30 @@
...
+configuration files.
 
 The interface is generated by `python maint/gen_binding_c.py`, executed
-inside the autogen.sh. It loads master config file, then loads per category
-custom config files and generate one file for each configured function in
-the src/binding/c folder, as well as necessary src/binding/c/Makefile.mk,
-src/binding/c/errnames.txt, and src/include/mpir_impl.h.
+inside of `autogen.sh`. It loads the master configuration file, then loads, per
+category, custom configuration files. One file is generated for each configured
+function in the `src/binding/c` folder. Additionally the it generates the
```
Comment (hzhou) [inline doc/wiki/notes/binding_c.md:16]: extra "the"

File: doc/wiki/notes/ch4_inline.md [hunk 3]
```diff
@@ -1,83 +1,86 @@
...
+sensitive code path, communication. For other functions like, `Init/Finalize`,
+`Comm/Win`, `Create/Destroy`, inlining is less beneficial.
+
+## Guidelines For Inlining
+
+1. PT2PT, Collective and RMA communication functions need to be inlined. This
+   also means any function that is used in this code path needs to be inlined
+   as well. For example, request creation or AM fallback for communication.
+
+2. Any function that is not described by the first rule should not be inlined.
```
Comment (hzhou) [inline doc/wiki/notes/ch4_inline.md:16]: Change "should not" to "need not to".

File: doc/wiki/notes/ch4_inline.md [hunk 4]
```diff
@@ -1,83 +1,86 @@
...
+
+3. For a given component some of its functions may need to be inlined
+   while others may not. For example, in `ch4r_buf.h/.c`, the creation and
+   destroy of `buf pool` is only used in `init/finalize` code. Therefore no
+   inlining is needed. The functions for obtaining and returning individual
+   buffers are used in the communication code path, and require inlining.
```
Comment (hzhou) [inline doc/wiki/notes/ch4_inline.md:22]: `buf pool` has been replaced by `GenQ`. This section no longer apply. Just remove.

File: doc/wiki/notes/ch4_inline.md [hunk 5]
```diff
@@ -1,83 +1,86 @@
...
+## Why is `MPIDI_NM_comm_get_gpid` moved to `ofi_proc.h`?
+
+This function is used in `MPIDIG RMA` code and needs to be inlined. Because it
+was the only function in `ofi_init.h` that should be inlined and was ugly to
+keep `ofi_init.h` just for this function, it was moved.
+
+It cannot be put in `ofi_impl.h` because `ofi_impl.h` is included in many
+`.c` files which in turn will create their own copy of `MPIDI_NM_comm_get_gpid`
+and conflict with each other. Therefore, it was put in `ofi_proc.h`.
+
```
Comment (hzhou) [inline doc/wiki/notes/ch4_inline.md:73]: The above two paragraphs explain a minor technical rationale or workaround. It is not critical for understanding and maintenance. Suggest removal.

File: doc/wiki/notes/ch4_inline.md [hunk 6]
```diff
@@ -1,83 +1,86 @@
...
+doing a `diff` between the original `.h` file and the new `.c` file. Try to
+maintain the relative location of the functions when moving them to the `.c`
+file so that it is easier to do a diff in the editor. Then something like this
+can be done:
+
+```
+vimdiff ofi_win.h ofi_win.c
+```
+
+to figure out what was changed during the conflict.
```
Comment (hzhou) [inline doc/wiki/notes/ch4_inline.md:86]: This section addresses a difficult right after uninlining work. It is no longer needed. Just remove.

File: doc/wiki/notes/ch4_namespace.md [hunk 7]
```diff
@@ -1,57 +1,50 @@
...
-  1. Component: a set of data structures and/or functions that related
-     to a particular functionality.
+This page discusses how namespaces should be used in CH4.
 
-= MPID_ namespace
-  This is the abstract device interface (ADI). Usually, these names are
-  linked to their corresponding implementations in the MPIDI_ namespace.
+Below is the format that namespaces will follow on this page:
+    1. `Component` - A set of data structures and/or functions that related
+       to a particular functionality.
```
Comment (hzhou) [inline doc/wiki/notes/ch4_namespace.md:7]: The original version is better. It explains a "term", not "format"

**Review by hzhou — COMMENTED** [review]
Some notes

File: doc/wiki/notes/ch4_inline.md (hunk 3 above)
Comment (travisk1154) [reply doc/wiki/notes/ch4_inline.md:16]: is "...does not need to be inlined." ok?

File: doc/wiki/notes/ch4_inline.md (hunk 4 above)
Comment (travisk1154) [reply doc/wiki/notes/ch4_inline.md:22]: `For a given component some of its functions may need to be inlined while others may not.`
 Should I keep this line and remove the example?

File: doc/wiki/Index.md (hunk 1 above)
Comment (hzhou) [reply doc/wiki/Index.md:38]: No. "Design Documentation" currently is a catch bag.

File: doc/wiki/notes/ch4_inline.md (hunk 3 above)
Comment (hzhou) [reply doc/wiki/notes/ch4_inline.md:16]: Sure.

File: doc/wiki/notes/ch4_inline.md (hunk 5 above)
Comment (travisk1154) [reply doc/wiki/notes/ch4_inline.md:73]: Removed.

File: doc/wiki/notes/ch4_inline.md (hunk 6 above)
Comment (travisk1154) [reply doc/wiki/notes/ch4_inline.md:86]: Removed.

File: doc/wiki/notes/ch4_inline.md (hunk 4 above)
Comment (hzhou) [reply doc/wiki/notes/ch4_inline.md:22]: I would say just remove the whole thing.

File: doc/wiki/notes/ch4_namespace.md (hunk 7 above)
Comment (travisk1154) [reply doc/wiki/notes/ch4_namespace.md:7]: Reverted.

File: doc/wiki/notes/ch4_inline.md (hunk 3 above)
Comment (travisk1154) [reply doc/wiki/notes/ch4_inline.md:16]: Fixed.

File: doc/wiki/notes/binding_c.md (hunk 2 above)
Comment (travisk1154) [reply doc/wiki/notes/binding_c.md:16]: Fixed.

File: doc/wiki/notes/ch4_inline.md (hunk 4 above)
Comment (travisk1154) [reply doc/wiki/notes/ch4_inline.md:22]: Removed.

File: doc/wiki/Index.md (hunk 1 above)
Comment (travisk1154) [reply doc/wiki/Index.md]: Removed this commit.

**Review by hzhou — APPROVED** [review]
LGTM

**Pull request merged**